Do not provide share link if (annotation or document) URL is not available on the web

lyzadanger opened the pull request:
There has been some (understandable) ongoing confusion around share links for single annotations or sets of annotations that pertain to a URI that is not available on the web. Specifically, it's confusing if one annotates a local PDF and then attempts to share a single annotation or a set of annotations on that PDF. Before these changes, the application would happily cook up a bouncer (sharing) link for you, but that link wouldn't do anything useful.

## With these changes...

*Note*: For convenience, the screenshots here show both share-single-annotation and share-a-page's-annotations open at the same time. While this isn't broken or impossible, it's probably slightly unusual.

These changes create a new UI path for both the share-single-annotation and share-a-page's-annotations components. When the URI is "shareable," there is no change from before, which looks like this:

<img width="469" alt="Screen Shot 2021-01-07 at 1 57 23 PM" src="https://user-images.githubusercontent.com/439947/103933196-f8b64500-50f0-11eb-9fa5-276e760fdd2e.png">

If the URI in question is not "shareable," the interfaces will look like this:

<img width="470" alt="Screen Shot 2021-01-07 at 2 02 19 PM" src="https://user-images.githubusercontent.com/439947/103933236-08ce2480-50f1-11eb-8ea9-6564a055ead8.png">

To support these changes, component changes were needed (duh), but a few other supporting changes were made as well to the `util/annotation-sharing` module, and some adjustments to `card` and `panel` molecules that I'll elaborate on with inline comments in this PR.

I manually tested that:

* When built into the local extension, local PDF files appropriately get the new, unshareable treatment.
* When built into the local extension, web-hosted PDF files still allow sharing correctly.

Fixes #2786

lyzadanger commented on src/sidebar/components/share-annotations-panel.js:
```diff
@@ -30,24 +31,24 @@ function ShareAnnotationsPanel({ analytics, toastMessenger }) {
   const copyShareLink = () => {
     try {
-      copyText(shareURI);
+      copyText(shareURI ?? '');
```
Appeasing the TS gods

robertknight replied:
Even though it is a bit verbose I would prefer to use a cast here (`/** @type {string} */ (shareURI)`). This avoids adding an untested path to the code and also makes it more obvious to future readers that this is purely to keep TS happy.

As an aside, this is one area where TypeScript's native syntax is much more succinct (just add an exclamation mark after the expression - `shareURI!`).

lyzadanger commented on src/sidebar/components/share-annotations-panel.js:
```diff
@@ -59,10 +60,10 @@ function ShareAnnotationsPanel({ analytics, toastMessenger }) {
         <div className="share-annotations-panel">
           <div className="share-annotations-panel__intro">
-            {focusedGroup.type === 'private' ? (
+            {focusedGroup?.type === 'private' ? (
```
TS again...

lyzadanger commented on src/sidebar/util/annotation-sharing.js:
```diff
@@ -1,47 +1,61 @@
+ * @param {string} groupID
+ * @return {string}
+ */
+export function getSharingLink(documentURI, groupID) {
```
This function's logic was moved from the `ShareAnnotationsPanel` component to here.

lyzadanger commented on src/sidebar/util/annotation-sharing.js:
```diff
@@ -1,47 +1,61 @@
 /**
- * Return any defined standalone URI for this `annotation`, preferably the
- * `incontext` URI, but fallback to `html` link if not present.
+ * Return the service-provided sharing URI for an annotation. Prefer the
```
Just some minor improvements and typing while I was in here.

lyzadanger commented on src/sidebar/util/annotation-sharing.js:
```diff
@@ -1,47 +1,61 @@
  */
-export function isShareable(annotation, settings) {
-  return !!(sharingEnabled(settings) && shareURI(annotation));
+export function sharingEnabled(settings) {
```
Note that we check config (i.e. call this function) to determine whether it should be allowed to share single annotations, but the logic for showing the "share all annotations on this page" is simply "is this first-party?" (tested in `AnnotationActionBar`).

lyzadanger commented on src/styles/mixins/molecules.scss:
```diff
@@ -20,9 +20,9 @@
  * hover/active intensified box shadow and accounts for "theme-clean" affordances.
  */
-@mixin card {
+@mixin card($rhythm: var.$layout-space) {
```
The updates to this module's mixins were because the `panel--compact` mixin was initially designed for the specific "design case" of the share-single-annotation small panel with the text-input field in it, instead of the more "generic" header-and-text layout that is used in the new UI path (not shareable). This fixes it such that the tighter padding and rhythm in the variant that has the input field is the _exception_ rather than the _rule_.

lyzadanger commented on src/styles/sidebar/components/annotation-share-control.scss:
```diff
@@ -25,9 +25,15 @@
+  // Override the pointer cursor that applies to the entire card
   cursor: default;
 
+  // Hide the bottom border on the panel's header if displaying
```
This is where the variant _with_ the sharing link overrides the generic  panel styling instead of vice-versa.

lyzadanger commented on src/sidebar/components/annotation-action-bar.js:
```diff
@@ -53,7 +53,7 @@ function AnnotationActionBar({
   const showFlagAction =
     !!userProfile.userid && userProfile.userid !== annotation.user;
-  const showShareAction = isShareable(annotation, settings);
+  const showShareAction = sharingEnabled(settings) && shareURI(annotation);
```
The overall refactor of `annotation-sharing` logic made it more coherent to have this logic in two discrete functions instead of one. There is some complexity because "is sharing allowed" is not a single question, but, kind of, four:

* Should we show the icon for sharing this page's annotations? The logic we use (unchanged) is "Are we first-party?" This is in `AnnotationActionBar` at present.
* Should we show the icon for sharing a single annotation? (This line here). The logic we use (unchanged) is: is there anything in the app's config that says we _shouldn't_ allow annotation sharing, and does this annotation have a valid single-annotation URL in its `links`?
* Which UI should we be providing within the "share all annotations on this page" panel, once it's open? (New). If the page's URI is shareable (i.e. web-accessible), we render the main, pre-existing UI that provides a bouncer link. If not, we render the new UI path explaining why the page's annotations are not shareable (with no link).
* Which UI should we be providing within the "share this annotation" panel, once it's open? (New). If the annotation's URI is shareable (i.e. web-accessible), we render the main, pre-existing UI that provides a bouncer link. If not, we render the new UI path explaining why this annotation is not shareable (with no link).

lyzadanger commented on src/sidebar/util/annotation-sharing.js:
```diff
@@ -1,47 +1,61 @@
-  }
-  return serviceConfig_.enableShareLinks;
+export function isShareableURI(uri) {
+  return /^http(s?):/.test(uri);
```
So, this is dead simple. Is it correct?

I opted not to instantiate a `URL` object here, though we certainly could.

robertknight reviewed (commented):
My initial thought after looking at this is that the text may be misleading. Where it says "the annotation can't be shared" what it means is "the annotation can't be shared _in context_". It is still possible (and useful to users) to share the content of the annotation itself and the discussion via a link to the annotation or a link to a notebook/activity page view that shows the annotations.

One possibility would be to show the same UI when the document URL is shareable as when it is not, but change the type of link (eg. show the `html` link for an annotation instead of the `incontext` one). However it may be confusing if the link does different things in different situations. Perhaps some explanatory text below the link may be enough to resolve that.

Regarding the code, I noted an instance of commented out code and had some recommendations about working around limitations in TypeScript's ability to infer when a value is non-null.

lyzadanger commented on src/sidebar/util/annotation-sharing.js:
```diff
@@ -1,47 +1,61 @@
+export function getSharingLink(documentURI, groupID) {
+  return `https://hyp.is/go?url=${encodeURIComponent(
+    documentURI
+  )}&group=${groupID}`;
```
YES.

lyzadanger commented on src/sidebar/components/test/share-annotations-panel-test.js:
```diff
@@ -132,30 +141,49 @@ describe('ShareAnnotationsPanel', () => {
+      // assert.equal(
+      //   inputEl.prop('value'),
+      //   'https://hyp.is/go?url=https%3A%2F%2Fwww.example.com&group=testprivate'
+      // );
```
Thanks for your patience on this. I've added it to my list of "stuff I sometimes overlook in PRs" to see if I can be more careful about this in the future. There's no straightforward way to lint for it, so it might just rely on me being a little more attentive!

lyzadanger commented on src/sidebar/components/test/share-annotations-panel-test.js:
```diff
@@ -38,6 +40,9 @@ describe('ShareAnnotationsPanel', () => {
       copyText: sinon.stub(),
     };
+
+    fakeGetSharingLink = sinon.stub().returns('bouncer-link');
```
That seems very reasonable. Good point.

lyzadanger commented:
@robertknight I've addressed your code feedback on my working branch. Given that there was a second iteration of how sharing single annotations is handled in these cases, I'm going to open a fresh PR (with your feedback incorporated) for easier navigation.
